pkg/operation/botanist/dns.go

```diff
@@ -15,329 +15,194 @@
-	"github.com/gardener/gardener/pkg/operation/terraformer"
-	"github.com/gardener/gardener/pkg/utils"
-	corev1 "k8s.io/api/core/v1"
+	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
+
+	dnsv1alpha1 "github.com/gardener/external-dns-management/pkg/apis/dns/v1alpha1"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var dnsChartPath = filepath.Join(common.ChartPath, "seed-dns")
+
+const (
+	// DNSPurposeInternal is a constant for a DNS record used for the internal domain name.
+	DNSPurposeInternal = "internal"
+	// DNSPurposeExternal is a constant for a DNS record used for the external domain name.
+	DNSPurposeExternal = "external"
 )
 
 // DeployInternalDomainDNSRecord deploys the DNS record for the internal cluster domain.
-func (b *Botanist) DeployInternalDomainDNSRecord() error {
-	return b.DeployDNSRecord(common.TerraformerPurposeInternalDNS, b.Shoot.InternalClusterDomain, b.APIServerAddress, true)
+func (b *Botanist) DeployInternalDomainDNSRecord(ctx context.Context) error {
+	if err := b.deployDNSProvider(ctx, DNSPurposeInternal, b.Garden.InternalDomain.Provider, b.Garden.InternalDomain.SecretData, b.Shoot.InternalClusterDomain); err != nil {
+		return err
+	}
+	if err := b.deployDNSEntry(ctx, DNSPurposeInternal, b.Shoot.InternalClusterDomain, b.APIServerAddress); err != nil {
+		return err
+	}
+	return b.deleteLegacyTerraformDNSResources(ctx, common.TerraformerPurposeInternalDNSDeprecated)
 }
 
 // DestroyInternalDomainDNSRecord destroys the DNS record for the internal cluster domain.
-func (b *Botanist) DestroyInternalDomainDNSRecord() error {
-	return b.DestroyDNSRecord(common.TerraformerPurposeInternalDNS, true)
+func (b *Botanist) DestroyInternalDomainDNSRecord(ctx context.Context) error {
+	if err := b.deleteDNSEntry(ctx, DNSPurposeInternal); err != nil {
+		return err
+	}
+	return b.deleteDNSProvider(ctx, DNSPurposeInternal)
 }
 
 // DeployExternalDomainDNSRecord deploys the DNS record for the external cluster domain.
-func (b *Botanist) DeployExternalDomainDNSRecord() error {
-	return b.DeployDNSRecord(common.TerraformerPurposeExternalDNS, *(b.Shoot.ExternalClusterDomain), b.Shoot.InternalClusterDomain, false)
+func (b *Botanist) DeployExternalDomainDNSRecord(ctx context.Context) error {
+	if b.Shoot.Info.Spec.DNS.Domain == nil || b.Shoot.ExternalClusterDomain == nil || strings.HasSuffix(*b.Shoot.ExternalClusterDomain, ".nip.io") {
+		return nil
+	}
+
+	if err := b.deployDNSProvider(ctx, DNSPurposeExternal, b.Shoot.ExternalDomain.Provider, b.Shoot.ExternalDomain.SecretData, *b.Shoot.Info.Spec.DNS.Domain); err != nil {
+		return err
+	}
+	if err := b.deployDNSEntry(ctx, DNSPurposeExternal, *b.Shoot.ExternalClusterDomain, b.Shoot.InternalClusterDomain); err != nil {
+		return err
+	}
+	return b.deleteLegacyTerraformDNSResources(ctx, common.TerraformerPurposeExternalDNSDeprecated)
```

Comment:
Why are you unsure? Can you point out concrete technical problems/uncertainties? We only delete the old Terraform resourcs if the DNS entry is "Ready", i.e. the responsible DNS controller has taken over responsibility. No need to keep the old Terraform logic from my point of view - unless there is a scenario where it might be necessary (which I don't see).

Reply:
I was referring to a case where `.spec.DNS.Domain` was changed, but then I realized that it was immutable.